Make reindex throttling dynamic

nik9000 opened the pull request:
This creates a rest end point that lets the user change the throttle of reindex. It takes care to reschedule the task if the user tries to speed up the request. This is important so that users undo throttle values that make reindex sleep forever. We also listen for cancelation and wake up if we are sleeping so that cancelled reindex requests that will sleep for a long time die quickly.

Finally, this adds a field to the task status that is "for how much longer will this request sleep?" I needed it debugging some stuff and see no reason to remove it.

nik9000 commented:
This totally needs docs. Also, I'm not sure if this is a good or bad REST location:
```
        controller.registerHandler(POST, "/_update_by_query/{taskId}/_rethrottle", this);
        controller.registerHandler(POST, "/_reindex/{taskId}/_rethrottle", this);
```

@imotov I think maybe this is a good one for you to review? Or @dakrone? I'm not sure. A tiny bit of task management stuff, mostly reindex stuff.

nik9000 commented:
@dakrone can you have a look at this one?

clintongormley commented:
Hmmm wondering if this should be more generic eg some way of sending messages to tasks via the task management API?

nik9000 commented:
> Hmmm wondering if this should be more generic eg some way of sending messages to tasks via the task management API?

The task management bit of the implementation is basically just that. But at the API level it isn't generic.

dakrone commented on modules/reindex/src/main/java/org/elasticsearch/index/reindex/AbstractAsyncBulkByScrollAction.java:
```diff
@@ -173,52 +173,60 @@ void onScrollResponse(TimeValue delay, SearchResponse searchResponse) {
         task.setTotal(total);
-        task.countThrottle(delay);
-        threadPool.schedule(delay, ThreadPool.Names.GENERIC, threadPool.getThreadContext().preserveContext(new AbstractRunnable() {
+        AbstractRunnable prepareBulkRequest = (AbstractRunnable) threadPool.getThreadContext().preserveContext(new AbstractRunnable() {
```
It was confusing to me that you call this `prepareBulkRequest` but also have a method named `prepareBulkRequest`, perhaps one of them can be renamed?

nik9000 replied:
Will rename.

dakrone commented on modules/reindex/src/main/java/org/elasticsearch/index/reindex/AbstractBaseReindexRestHandler.java:
```diff
@@ -43,6 +43,21 @@
+        if (requestsPerSecond == null) {
+            return null;
+        }
+        if ("unlimited".equals(requestsPerSecond)) {
```
I know we sometimes use `-1` to indicate disabling something (like refresh interval), does this need to be parsed here as well?

nik9000 replied:
I can allow -1 to mean unlimited as well.

dakrone replied:
Cool, I think we should allow the -1 to it conceptually makes sense :)

dakrone commented on modules/reindex/src/main/java/org/elasticsearch/index/reindex/BulkByScrollTask.java:
```diff
@@ -50,15 +58,39 @@
+     * style variables but there isn't an AtomicFloat so we just use a volatile.
+     */
+    private volatile float requestsPerSecond;
+    private final AtomicReference<DelayedPrepareBulkRequest> delayedPrepareBulkRequest = new AtomicReference<>();
```
Same here with naming, we have `delayedPrepareBulkRequest` and `delayPrepareBulkRequest` which are confusing when vars and functions are named so similarly

dakrone commented on modules/reindex/src/main/java/org/elasticsearch/index/reindex/BulkByScrollTask.java:
```diff
@@ -298,10 +356,106 @@ void countRetry() {
+                @Override
+                protected void doRun() throws Exception {
+                    // Paranoia to prevent furiously rethrottling from running the command multiple times. Without this we totally can.
+                    if (false == oneTime.getAndSet(true)) {
```
This might read better if it were renamed to `hasRun` initialized with false and then it could be

```java
if (hasRun.compareAndSet(false, true)) {
   prepareBulkRequest.run();
}
```

nik9000 replied:
Much nicer.

dakrone commented on modules/reindex/src/main/java/org/elasticsearch/index/reindex/RethrottleRequest.java:
```diff
@@ -0,0 +1,68 @@
+ */
+public class RethrottleRequest extends BaseTasksRequest<RethrottleRequest> {
+    /**
+     * The throttle to apply to all matching requests in sub-requests per second. 0 means set no throttle and that is the default.
```
As I mentioned above, I do think we should support both 0 and -1 for no throttle to be consistent with our other "disabling" APIs

dakrone commented on rest-api-spec/src/main/resources/rest-api-spec/api/reindex.rethrottle.json:
```diff
@@ -0,0 +1,24 @@
+      "params": {
+        "requests_per_second": {
+          "type": "float",
+          "default": 0,
```
Does this then mean that doing `curl -XPOST 'localhost:9200/_reindex/1234/_rethrottle?requests_per_second'` will set the task to be unlimited requests per sec?

nik9000 replied:
I'm not sure what that'll do. I wrote the default that way to mean "if you don't send the parameter at all it'll default to unlimited." I've not tried sending the parameter without a value - probably an exception parsing the float.

nik9000 replied:
I can make whatever you think is right work. I don't really have an opinion. I can also tell the user it is an error to set the parameter to empty string.

dakrone replied:
Yeah, it'll probably freak out parsing it as an empty string, perhaps we should check `Strings.hasText(...)` and throw a nicer exception if not?

nik9000 commented:
Thanks for the review @dakrone ! I pushed another commit that addresses your points - except for the one about the default value which I left a question on inline.

dakrone commented on core/src/main/java/org/elasticsearch/tasks/CancellableTask.java:
```diff
@@ -65,4 +66,10 @@ public boolean isCancelled() {
     }
+
+    /**
+     * Called when the task in cancelled so that it can take any actions that it has to take.
```
in -> is

Might be good to specify also whether it's called *before* cancellation, or *after*

nik9000 replied:
e743c2b

dakrone commented on modules/reindex/src/main/java/org/elasticsearch/index/reindex/AbstractAsyncBulkByScrollAction.java:
```diff
@@ -173,52 +173,61 @@ void onScrollResponse(TimeValue delay, SearchResponse searchResponse) {
+    }
+
+    /**
+     * Prepare the bulk request. Called on the generic thread pool after some preflight checks have been done one the SearchResponse and any
```
Just curious, but why is this on the generic thread pool instead of an action specific thread pool?

nik9000 replied:
I was going to type a comment but I realized that I'd be better off just pushing 577fc68.

dakrone commented on modules/reindex/src/main/java/org/elasticsearch/index/reindex/AbstractBaseReindexRestHandler.java:
```diff
@@ -61,7 +76,11 @@ protected AbstractBaseReindexRestHandler(Settings settings, Client client,
     protected void execute(RestRequest request, Request internalRequest, RestChannel channel) throws IOException {
-        internalRequest.setRequestsPerSecond(request.paramAsFloat("requests_per_second", internalRequest.getRequestsPerSecond()));
+        Float requestsPerSecond = parseRequestsPerSecond(request);
+        if (requestsPerSecond != null) {
```
Totally up to you (whether you want to or not), but maybe `setRequestsPerSecond` should handle the null case so you don't have to worry about checking it post-parse? What do you think? (I'm fine either way)

nik9000 replied:
I guess I like it better this way because null doesn't make sense in the context of the other caller. Does that make sense?

dakrone commented:
I left a few more comments (minor ones), but otherwise LGTM, thanks @nik9000

nik9000 commented:
> I left a few more comments (minor ones), but otherwise LGTM, thanks @nik9000

Thank you! I've pushed some more javadoc in response to your comments.
